model/src/main/kotlin/utils/FindingsMatcher.kt

```diff
@@ -292,62 +291,48 @@ internal fun associateLicensesWithExceptions(license: SpdxExpression): SpdxExpre
-            it is SpdxSimpleExpression -> simpleLicenses += it
-            else -> handledLicenses += it
-        }
+    val (standAloneExceptions, licenses) = license.children.partition {
+        it is SpdxSimpleExpression && SpdxLicenseException.forId(it.toString()) != null
     }
 
-    val i = remainingExceptions.iterator()
+    val handledExceptions = mutableSetOf<String>()
 
-    while (i.hasNext()) {
-        val exception = i.next()
-        val exceptionString = exception.toString()
+    val associatedLicenses = licenses.mapTo(mutableSetOf()) { childLicense ->
+        when (childLicense) {
+            is SpdxCompoundExpression -> associateLicensesWithExceptions(childLicense)
 
-        // Determine all licenses the exception is applicable to.
-        val applicableLicenses = SpdxLicenseException.mapping[exceptionString].orEmpty().mapTo(mutableSetOf()) {
-            SpdxLicenseIdExpression(it.id)
-        }
+            is SpdxSimpleExpression -> {
+                val licenseId = childLicense.toString()
 
-        // Associate all remaining licenses that are applicable with the exception and remove the exception.
-        val licenses = simpleLicenses.intersect(applicableLicenses)
-        if (licenses.isEmpty()) continue
+                val validLicenseExceptionCombinations = standAloneExceptions.mapNotNull { exception ->
+                    val exceptionId = exception.toString()
+                    val applicableLicenseId = SpdxLicenseException.mapping[exceptionId].orEmpty().map { it.id }
 
-        licenses.forEach {
-            handledLicenses += SpdxLicenseWithExceptionExpression(it, exceptionString)
-        }
+                    SpdxLicenseWithExceptionExpression(childLicense, exceptionId).takeIf {
+                        licenseId in applicableLicenseId
+                    }
+                }
 
-        associatedLicenses += licenses
-        i.remove()
-    }
+                handledExceptions += validLicenseExceptionCombinations.map { it.exception }
+                validLicenseExceptionCombinations.reduceOrNull(SpdxExpression::and) ?: childLicense
+            }
 
-    handledLicenses += simpleLicenses - associatedLicenses
+            else -> childLicense
+        }
+    }
 
     // Associate remaining "orphan" exceptions with "NOASSERTION" to turn them into valid SPDX expressions.
-    handledLicenses += remainingExceptions.map {
-        SpdxLicenseWithExceptionExpression(SpdxLicenseIdExpression(SpdxConstants.NOASSERTION), it.toString())
+    val orphanExceptions = standAloneExceptions.map { it.toString() } - handledExceptions
+
+    orphanExceptions.mapTo(associatedLicenses) { exceptionId ->
+        SpdxLicenseWithExceptionExpression(SpdxLicenseIdExpression(SpdxConstants.NOASSERTION), exceptionId)
     }
 
     // Recreate the compound AND-expression from the associated licenses.
-    return handledLicenses.reduce(SpdxExpression::and)
+    return associatedLicenses.reduce(SpdxExpression::and)
```

Comment:
See comment above regarding the use of `reduce`.

Reply:
Using reduce, though, has the advantage of transparently dealing with `associatedLicenses.size < 2` cases...